vortex-sampling-compressor/src/compressors/mod.rs

```diff
@@ -201,7 +223,36 @@ impl<'a> CompressedArray<'a> {
             let _ = stats.compute_uncompressed_size_in_bytes();
             array.inherit_statistics(stats);
         }
-        Self { array, path }
+        let compressed = Self { array, path };
+        compressed.validate();
+        compressed
+    }
+
+    fn validate(&self) {
+        self.validate_children(self.path.as_ref(), &self.array)
+    }
+
+    fn validate_children(&self, path: Option<&CompressionTree>, array: &ArrayData) {
+        if let Some(path) = path.as_ref() {
+            path.children
+                .iter()
+                .zip_longest(array.children().iter())
+                .for_each(|pair| match pair {
+                    EitherOrBoth::Both(Some(child_tree), child_array) => {
+                        self.validate_children(Some(child_tree), child_array);
+                    }
+                    EitherOrBoth::Left(Some(child_tree)) => {
+                        vortex_panic!(
+                            "Child tree without child array!!\nroot tree: {}\nroot array: {}\nlocal tree: {path}\nlocal array: {}\nproblematic child_tree: {child_tree}",
```

Comment:
I chose to panic rather than making this fallible because this is definitely programmer error (and also, we explore ~all possible permutations of encoding trees in the unit test suite)